tap_xero/__init__.py

```diff
@@ -88,23 +76,34 @@ def init_credentials(config):
                 config = credentials.refresh(config)
             except Exception as ex:
                 raise BadCredsException(BAD_CREDS_MESSAGE) from ex
-
     return config
 
 
-def sync(config, state, catalog):
-    init_credentials(config)
-    currently_syncing = state.get("currently_syncing")
+def load_and_write_schema(stream):
+    singer.write_schema(
+        stream.tap_stream_id,
+        load_schema(stream.tap_stream_id),
+        stream.pk_fields,
+    )
+
+
+def sync(ctx):
+    init_credentials(ctx.config)
+    currently_syncing = ctx.state.get("currently_syncing")
     start_idx = streams_.all_stream_ids.index(currently_syncing) \
         if currently_syncing else 0
-    stream_ids_to_sync = [c.tap_stream_id for c in catalog.streams
-                          if c.is_selected()]
-    for stream in streams_.all_streams[start_idx:]:
-        if stream.tap_stream_id not in stream_ids_to_sync:
-            continue
-        run_stream(config, state, stream)
-    state["currently_syncing"] = None
-    singer.write_state(state)
+    stream_ids_to_sync = [cs.tap_stream_id for cs in ctx.catalog.streams
+                          if cs.is_selected()]
+    streams = [s for s in streams_.all_streams[start_idx:]
+               if s.tap_stream_id in stream_ids_to_sync]
+    for stream in streams:
+        load_and_write_schema(stream)
+        ctx.state["currently_syncing"] = stream.tap_stream_id
+        ctx.write_state()
+        stream.sync(ctx)
```

Comment:
After this line, ctx currently_syncing should be set to None.  That or lines 100 and 101 should be swapped.  Otherwise when writing a streams schema, the ctx will point to the previous stream.

Reply:
That makes sense. I will swap the lines.